saq/queue/postgres.py

```diff
@@ -0,0 +1,608 @@
+
+    async def finish(
+        self,
+        job: Job,
+        status: Status,
+        *,
+        result: t.Any = None,
+        error: str | None = None,
+    ) -> None:
+        self._update_job_for_finish(job, status, result=result, error=error)
+        key = job.key
+
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            if job.ttl >= 0:
+                await cursor.execute(
+                    SQL(
+                        """
+                    UPDATE {jobs_table} SET status = %(status)s, job = %(job)s, ttl = %(ttl)s
+                    WHERE key = %(key)s
+                    """
+                    ).format(jobs_table=self.jobs_table),
+                    {
+                        "status": status,
+                        "job": self.serialize(job),
+                        "key": key,
+                        "ttl": seconds(now()) + job.ttl,
+                    },
+                )
+            else:
+                await cursor.execute(
+                    SQL(
+                        """
+                    DELETE FROM {jobs_table}
+                    WHERE key = %(key)s
+                    """
+                    ).format(jobs_table=self.jobs_table),
+                    {"key": key},
+                )
+        await self._release_job(key)
+
+        self._update_stats(status)
+
+        await self.notify(job)
+        logger.info("Finished %s", job.info(logger.isEnabledFor(logging.DEBUG)))
+
+    async def dequeue(self, timeout: float = 0) -> Job | None:
+        """Wait on `self.cond` to dequeue.
+
+        Retries indefinitely until a job is available or times out.
+        """
+        self.waiting += 1
+        async with self.cond:
+            self.cond.notify(1)
+
+        try:
+            return await asyncio.wait_for(self.queue.get(), timeout or None)
+        except asyncio.exceptions.TimeoutError:
+            return None
+        finally:
+            self.waiting -= 1
+
+    async def wait_for_job(self) -> None:
+        while True:
+            async with self.cond:
+                await self.cond.wait()
+
+            for job in await self._dequeue():
+                await self.queue.put(job)
+
+    async def enqueue(self, job_or_func: str | Job, **kwargs: t.Any) -> Job | None:
+        job = self._create_job_for_enqueue(job_or_func, **kwargs)
+
+        await self._before_enqueue(job)
+
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                SQL(
+                    """
+                INSERT INTO {jobs_table} (key, job, queue, status, scheduled)
+                VALUES (%(key)s, %(job)s, %(queue)s, %(status)s, %(scheduled)s)
+                ON CONFLICT (key) DO UPDATE
+                SET job = %(job)s, queue = %(queue)s, status = %(status)s, scheduled = %(scheduled)s
+                WHERE {jobs_table}.status NOT IN ('queued', 'aborting')
```

Comment:
so we won't update jobs that are queued or aborting, but we will update them if they are active???

Reply:
Good point. I think we should only update, aka re-queue, jobs that are done, i.e. finished, failed, or aborted. What do you think?

Reply:
TERMINAL_STATUSES right?

Reply:
Yes, TERMINAL_STATUSES